oil_lang/expr_eval.py

```diff
@@ -176,10 +183,156 @@ def _ValueToPyObj(val):
+
+    elif case(value_e.Str):
+      left = cast(value.Str, UP_left)
+      right = cast(value.Str, UP_right)
+      return left.s == right.s
+
+    elif case(value_e.MaybeStrArray):
+      left = cast(value.MaybeStrArray, UP_left)
+      right = cast(value.MaybeStrArray, UP_right)
+      if len(left.strs) != len(right.strs):
+        return False
+
+      for i in xrange(0, len(left.strs)):
+        if left.strs[i] != right.strs[i]:
+          return False
+
+      return True
+
+    elif case(value_e.List):
+      left = cast(value.List, UP_left)
+      right = cast(value.List, UP_right)
+      if len(left.items) != len(right.items):
+        return False
+
+      for i in xrange(0, len(left.items)):
+        if not _ValuesEqual(left.items[i], right.items[i]):
+          return False
+
+      return True
+
+    elif case(value_e.AssocArray):
+      left = cast(value.Dict, UP_left)
+      right = cast(value.Dict, UP_right)
+      if len(left.d) != len(right.d):
+        return False
+
+      for k in left.d.keys():
+        if not k in right.d or right.d[k] != left.d[k]:
+          return False
+
+      return True
+
+    elif case(value_e.Dict):
+      left = cast(value.Dict, UP_left)
+      right = cast(value.Dict, UP_right)
+      if len(left.d) != len(right.d):
+        return False
+
+      for k in left.d.keys():
+        if not k in right.d or not _ValuesEqual(right.d[k], left.d[k]):
+          return False
+
+      return True
+
+  raise NotImplementedError()
+
+
+def _ValueContains(needle, haystack):
+  # type: (value_t, value_t) -> bool
+  """ haystack must be a collection type. """
+
+  UP_needle = needle
+  UP_haystack = haystack
+  with tagswitch(haystack) as case:
+    if case(value_e.List):
+      haystack = cast(value.List, UP_haystack)
+      for item in haystack.items:
+        if _ValuesEqual(item, needle):
+          return True
+
+      return False
+
+    elif case(value_e.MaybeStrArray):
+      haystack = cast(value.MaybeStrArray, UP_haystack)
+      if needle.tag() != value_e.Str:
+        raise error.InvalidType('Expected Str', loc.Missing())
+
+      needle = cast(value.Str, UP_needle)
+      for s in haystack.strs:
+        if s == needle.s:
+          return True
+
+      return False
+
+    elif case(value_e.Dict):
+      haystack = cast(value.Dict, UP_haystack)
+      if needle.tag() != value_e.Str:
+        raise error.InvalidType('Expected Str', loc.Missing())
+
+      needle = cast(value.Str, UP_needle)
+      return needle.s in haystack.d
+
+    elif case(value_e.AssocArray):
+      haystack = cast(value.AssocArray, UP_haystack)
+      if needle.tag() != value_e.Str:
+        raise error.InvalidType('Expected Str', loc.Missing())
+
+      needle = cast(value.Str, UP_needle)
+      return needle.s in haystack.d
+
+    else:
+      raise error.InvalidType('Expected List or Dict', loc.Missing())
+
+  return False
   
 
+def AsPosixEre(eggex):
+  # type: (value.Eggex) -> str
+  if len(eggex.as_ere):
```

Comment:
I think this is clearer as `eggex.as_ere is not None`, i.e. it's set if it's not None

And `core/runtime.asdl` should have `str? as_ere`

Similar to oil_lang/objects.py (can we delete that?)

Not sure if it actually comes up, but  in theory you can have an empty eggex that translates to an empty regex ?  (probably should have a test case for that)